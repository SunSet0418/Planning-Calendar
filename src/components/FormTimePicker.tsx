import * as React from 'react';
import { TimePicker } from 'antd';
import { TimePickerProps } from 'antd/lib/time-picker';
import 'antd/lib/time-picker/style/css';
import { Moment } from 'moment';

interface OwnProps {
    name: string,
    defaultValue: Moment
}

type Props = OwnProps & TimePickerProps

export const FormTimePicker: React.FC<Props> = (props) => {
    const { name, defaultValue, ...pickerProps } = props;
    const [time, setTime] = React.useState(defaultValue.toString());
    const [open, setOpen] = React.useState(false);
    const handleChange = (moment: Moment, timeString: string) => {
        setOpen(false);
        setTime(timeString);
        if(props.onChange) {
            props.onChange(moment, timeString);
        }
    }
    return (
        <>
            <input name={name} type="number" hidden={true} value={time} />
            <TimePicker {...pickerProps}
                open={open}
                onOpenChange={setOpen}
                onChange={handleChange} />
        </>
    )
}