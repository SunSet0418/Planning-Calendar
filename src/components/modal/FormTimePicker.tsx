import * as React from 'react';
import styled from 'styled-components';
import {Dropdown, Menu, Button} from 'antd';
import 'antd/lib/dropdown/style/css';
import 'antd/lib/menu/style/css';
import { range } from 'lodash';
import { ClickParam } from 'antd/lib/menu';

const Label = styled.span`
    margin-left: 10px;
`

const TimePickerMenu = styled(Menu)`
    max-height: 200px;
    overflow: scroll;
`;

const TimePickerButton = styled(Button)`
    text-align: center;
    width: 40px;
    margin-left: 6px;
    padding: 0;
`

interface Props {
    name: string,
    value?: number,
    onChange?: (hour: number) => void,
    disableHours?: number[]
}

export const FormTimePicker: React.FC<Props> = ({name, value = 0, onChange, disableHours = []}) => {
    const [time, setTime] = React.useState(value);
    const handleTextChange = (e: React.ChangeEvent<HTMLInputElement>) => setTime(Number(e.target.value));
    const handleChange = (e: ClickParam) => {
        const hour = Number(e.key);
        setTime(hour);
        if(onChange) {
            onChange(hour);
        }
    }
    const timeRange = range(0, 25).filter(h => !disableHours.includes(h));
    const menu = (
        <TimePickerMenu onClick={handleChange} >
            {timeRange.map(hour => <Menu.Item key={hour}>{hour}</Menu.Item>)}
        </TimePickerMenu>
    )
    React.useEffect(() => setTime(value), [value]);
    return (
        <>
            <Label>{time > 12 ? '오후' : '오전'}</Label>
            <input name={name} type="string" hidden={true} value={time} onChange={handleTextChange} />
            <Dropdown overlay={menu} trigger={['click']}>
                <TimePickerButton>
                    {time}
                </TimePickerButton>
            </Dropdown>
        </>
    )
}