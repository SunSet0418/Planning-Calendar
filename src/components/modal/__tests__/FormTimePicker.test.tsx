import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as moment from 'moment';
import { TimePicker } from 'antd';
import { FormTimePicker } from '../FormTimePicker';

describe('<FormTimePicker />', () => {
    const testMoment = moment.parseZone('2014-02-27T00:00:00Z');
    let component = null;

    it('renders correctly', () => {
        component = Enzyme.shallow(<FormTimePicker name="form-name" defaultValue={testMoment}/>);
    });

    it('should match snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should render input, TimePicker', () => {
        expect(component.find('input').length).toBe(1);
        expect(component.find('input').props().name).toBe("form-name");
        expect(component.find(TimePicker).length).toBe(1);
    });

});