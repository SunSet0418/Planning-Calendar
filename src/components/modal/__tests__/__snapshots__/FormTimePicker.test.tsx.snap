// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormTimePicker /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    defaultValue={"2014-02-27T00:00:00.000Z"}
    name="form-name"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <input
          hidden={true}
          name="form-name"
          onChange={[Function]}
          type="string"
          value=""
/>,
        <TimePicker
          align={
                    Object {
                              "offset": Array [
                                0,
                                -2,
                              ],
                            }
          }
          defaultValue={"2014-02-27T00:00:00.000Z"}
          disabled={false}
          disabledHours={undefined}
          disabledMinutes={undefined}
          disabledSeconds={undefined}
          focusOnOpen={true}
          hideDisabledOptions={false}
          onChange={[Function]}
          onOpenChange={[Function]}
          open={false}
          placement="bottomLeft"
          transitionName="slide-up"
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "hidden": true,
          "name": "form-name",
          "onChange": [Function],
          "type": "string",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "align": Object {
            "offset": Array [
              0,
              -2,
            ],
          },
          "defaultValue": "2014-02-27T00:00:00.000Z",
          "disabled": false,
          "disabledHours": undefined,
          "disabledMinutes": undefined,
          "disabledSeconds": undefined,
          "focusOnOpen": true,
          "hideDisabledOptions": false,
          "onChange": [Function],
          "onOpenChange": [Function],
          "open": false,
          "placement": "bottomLeft",
          "transitionName": "slide-up",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <input
            hidden={true}
            name="form-name"
            onChange={[Function]}
            type="string"
            value=""
/>,
          <TimePicker
            align={
                        Object {
                                    "offset": Array [
                                      0,
                                      -2,
                                    ],
                                  }
            }
            defaultValue={"2014-02-27T00:00:00.000Z"}
            disabled={false}
            disabledHours={undefined}
            disabledMinutes={undefined}
            disabledSeconds={undefined}
            focusOnOpen={true}
            hideDisabledOptions={false}
            onChange={[Function]}
            onOpenChange={[Function]}
            open={false}
            placement="bottomLeft"
            transitionName="slide-up"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "hidden": true,
            "name": "form-name",
            "onChange": [Function],
            "type": "string",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "align": Object {
              "offset": Array [
                0,
                -2,
              ],
            },
            "defaultValue": "2014-02-27T00:00:00.000Z",
            "disabled": false,
            "disabledHours": undefined,
            "disabledMinutes": undefined,
            "disabledSeconds": undefined,
            "focusOnOpen": true,
            "hideDisabledOptions": false,
            "onChange": [Function],
            "onOpenChange": [Function],
            "open": false,
            "placement": "bottomLeft",
            "transitionName": "slide-up",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
